### PR TITLE
[SYCL][ESIMD] Randomizing inputs

### PR DESCRIPTION
**Pull request by dongkyunahn-intel**
- accessor_load_store.cpp is modified for randomized inputs

### CODE REVIEW

**Review by vladimirlaz — CHANGES_REQUESTED** [review]
It looks like we add instability to the test result by  adding randomization to the test. It is better to have stable data and stable results.

File: SYCL/ESIMD/esimd_test_utils.hpp [hunk 1]
```diff
@@ -461,4 +462,28 @@ inline void iterate_ops(OpSeq<OpClass, Ops...> ops, F f) {
...
+  explicit Generator() : _gen(std::random_device{}()), _dist() {}
+  Generator(T ub) : _gen(std::random_device{}()), _dist(0, ub) {}
+  Generator(T lb, T ub) : _gen(std::random_device{}()), _dist(lb, ub) {}
+
+  T operator()() { return static_cast<T>(_dist(_gen)); }
+
+private:
+  using dist_t = typename std::conditional<
+      std::is_integral<T>::value,
+      typename std::conditional<std::is_same<T, char>::value ||
```
Comment (lsatanov) [inline SYCL/ESIMD/esimd_test_utils.hpp:478]: Isn't this line redundant?: char is either unsigned or signed, both are mentioned below

File: SYCL/ESIMD/esimd_test_utils.hpp [hunk 2]
```diff
@@ -461,4 +462,28 @@ inline void iterate_ops(OpSeq<OpClass, Ops...> ops, F f) {
...
+                                    std::is_same<T, signed char>::value ||
+                                    std::is_same<T, unsigned char>::value,
+                                std::uniform_int_distribution<int>,
+                                std::uniform_int_distribution<T>>::type,
+      std::uniform_real_distribution<T>>::type;
```
Comment (lsatanov) [inline SYCL/ESIMD/esimd_test_utils.hpp:483]: The distribution type is not changeable?

File: SYCL/ESIMD/esimd_test_utils.hpp (hunk 1 above)
Comment (lsatanov) [reply SYCL/ESIMD/esimd_test_utils.hpp:478]: Question: if this class is supposed to be used  just with chars, why this check attached to distribution type definition, not the enclosing class?

File: SYCL/ESIMD/esimd_test_utils.hpp [hunk 3]
```diff
@@ -461,4 +462,28 @@ inline void iterate_ops(OpSeq<OpClass, Ops...> ops, F f) {
   ConstexprForLoop<0, sizeof...(Ops)>::unroll(act);
 }
 
+/* Randomized input generator */
+namespace input_gen {
+template <typename T> class Generator {
```
Comment (lsatanov) [inline SYCL/ESIMD/esimd_test_utils.hpp:467]: Question: seed-ing functionality is not required?

File: SYCL/ESIMD/esimd_test_utils.hpp [hunk 4]
```diff
@@ -461,4 +462,28 @@ inline void iterate_ops(OpSeq<OpClass, Ops...> ops, F f) {
...
+  using dist_t = typename std::conditional<
+      std::is_integral<T>::value,
+      typename std::conditional<std::is_same<T, char>::value ||
+                                    std::is_same<T, signed char>::value ||
+                                    std::is_same<T, unsigned char>::value,
+                                std::uniform_int_distribution<int>,
+                                std::uniform_int_distribution<T>>::type,
+      std::uniform_real_distribution<T>>::type;
+
+  std::mt19937_64 _gen;
```
Comment (lsatanov) [inline SYCL/ESIMD/esimd_test_utils.hpp:485]: Question: custom seed-ing functionality is not required?

File: SYCL/ESIMD/accessor_load_store.cpp [hunk 5]
```diff
@@ -48,10 +49,15 @@ template <typename T> struct char_to_int {
...
   std::cout << "Testing T=" << typeid(T).name() << "...\n";
+  esimd_test::input_gen::Generator<T> gen(
+      std::numeric_limits<T>::min() / (T)size,
+      std::numeric_limits<T>::max() / (T)size);
   T *A = new T[size];
+  T *G = new T[size];
 
   for (unsigned i = 0; i < size; ++i) {
-    A[i] = (T)i;
+    A[i] = gen();
```
Comment (lsatanov) [inline SYCL/ESIMD/accessor_load_store.cpp:59]: No seed-ing for reproducibility?

File: SYCL/ESIMD/accessor_load_store.cpp [hunk 6]
```diff
@@ -48,10 +49,15 @@ template <typename T> struct char_to_int {
 
 template <typename T> bool test(queue q, size_t size) {
   std::cout << "Testing T=" << typeid(T).name() << "...\n";
+  esimd_test::input_gen::Generator<T> gen(
```
Comment (lsatanov) [inline SYCL/ESIMD/accessor_load_store.cpp:52]: No custom seed-ing?

**Comment by lsatanov** [comment]
> It looks like we add instability to the test result by adding randomization to the test. It is better to have stable data and stable results.

Hi, if I understand it correctly, the generator being used is pseudo-random  one and the default seed is fixed across standard library implementations (5489u) [C++ standard, 29.6.3.2 Class template mersenne_twister_engine ]. So the sequence it produces is stable (can be also godbolt-ed https://godbolt.org/z/8Mbezxbc3).

**Comment by dongkyunahn-intel** [comment]
> Hi, if I understand it correctly, the generator being used is pseudo-random one and the default seed is fixed across standard library implementations (5489u) [C++ standard, 29.6.3.2 Class template mersenne_twister_engine ]. So the sequence it produces is stable (can be also godbolt-ed https://godbolt.org/z/8Mbezxbc3).

I think what Vladimir meant was 'predictable' results with known input values.

File: SYCL/ESIMD/esimd_test_utils.hpp (hunk 1 above)
Comment (dongkyunahn-intel) [reply SYCL/ESIMD/esimd_test_utils.hpp:478]: I did not noticed while I was importing this code from our internal validation codes. Will update.

File: SYCL/ESIMD/accessor_load_store.cpp (hunk 6 above)
Comment (dongkyunahn-intel) [reply SYCL/ESIMD/accessor_load_store.cpp:52]: I wanted usage of this generator as simple as possible. Maybe custom seeding can be added later.

File: SYCL/ESIMD/esimd_test_utils.hpp (hunk 2 above)
Comment (dongkyunahn-intel) [reply SYCL/ESIMD/esimd_test_utils.hpp:483]: Like my mention above, I tried to make this generator as simple as possible while distribution is not considered in other samples using `rand()/srand()`.

File: SYCL/ESIMD/esimd_test_utils.hpp (hunk 4 above)
Comment (dongkyunahn-intel) [reply SYCL/ESIMD/esimd_test_utils.hpp:485]: Maybe we can added custom seeding later.

**Comment by dongkyunahn-intel** [comment]
> It looks like we add instability to the test result by adding randomization to the test. It is better to have stable data and stable results.

(Provided that what you meant 'predictable/known data&result' with 'stable data', ) The reason why I used accessor_load_store.cpp as example is that I noticed that negative values are never used in the test while I was debugging failure. I'll revert this sample and apply example usage in other sample(s).

**Comment by dongkyunahn-intel** [comment]
There was a comment from a reviewer that randomized inputs could make debugging hard. Closing this PR.

**Comment by lsatanov** [comment]
Nevertheless, 
generated != randomized
And in this case it is reproducibly generated, isn't it?

**Pull request closed without merging**